### DISCUSSION TITLE
open3d / embree seems to return nothing when ray hits a vertex exactly?

### DISCUSSION THREAD

**Discussion by pwais** [isl-org/Open3D#6278] [Q&A]
dear o3d peeps,

i tried raycasting using code very similar to the demo and plotting the result as is done here: https://github.com/isl-org/Open3D/blob/6ddbcd5c9b8bf0b496e4151c7d7766af09e3dba7/examples/python/geometry/ray_casting_to_image.py#L37

in my scene, I have rays that actually exactly correspond to vertices of the input mesh, and thus I'd expect the embree / open3d raycaster to return vertex hits.  but the actual output appears to be a lot of misses / 0 depth returns:

![Screen Shot 2023-07-22 at 5 58 35 PM](https://github.com/isl-org/Open3D/assets/218145/6bfd9672-4cf7-4796-a4c8-8c4336e315ff)
![Screen Shot 2023-07-22 at 5 58 49 PM](https://github.com/isl-org/Open3D/assets/218145/97a300e1-7b2c-40b1-9498-ea0ff106e4f6)

When I offset the entire vertices mesh by epsilon, then I get ray-triangle intersections that make sense:

![Screen Shot 2023-07-22 at 6 00 04 PM](https://github.com/isl-org/Open3D/assets/218145/f97fcd78-004d-48a1-8220-4ccc3cae51b1)
![Screen Shot 2023-07-22 at 6 00 11 PM](https://github.com/isl-org/Open3D/assets/218145/ae1d3c7f-db85-4181-b86f-7beaa76f36a3)

Looking at the source code and embree, it's actually not really clear... what embree is supposed to return if a ray hits a vertex exactly?  

It's worth noting that Open3D uses `rtcIntersect1M()` https://github.com/isl-org/Open3D/blob/6ddbcd5c9b8bf0b496e4151c7d7766af09e3dba7/cpp/open3d/t/geometry/RaycastingScene.cpp#L318C13-L318C27

which is supposed to be deprecated and removed in the most recent version of embree: https://github.com/embree/embree/blob/0fcb306c9176221219dd15e27fe0527ed334948f/README.md?plain=1#L1383

I wonder if using the new API will result in well-defined output for when a ray intersects a vertex?

**Comment by benjaminum** [comment]
Hi @pwais,
Hitting exactly a vertex is a corner case. I wonder why there are so many direct hits in your case. 
Are the rays defined in a specific way that makes this more likely?

**Comment by pwais** [comment]
yes by construction there are lots of vertex hits.  oddly I see the issue go away after I built open3d from source.  But Dr Woop says actually I should not see so many issues on default dembree.  I will investigate further.
